### DISCUSSION TITLE
Potential of using ContainerLab as a library

### DISCUSSION THREAD

**Discussion by ldelossa** [srl-labs/containerlab#1905] [Q&A]
Hello, 

I'm curious about whether the maintainers of ContainerLab see it feasible to use ContainerLab as a library. 

I work on Cilium OSS and having the ability to utilize ContainerLab to configure network topologies would benefit us greatly. 

This is just a "feeler" on whether the authors think the code base is modular enough to use the bulk of the features in an embedded fashion. 

For instance, lets say Cilium (or its testing framework) wanted to define a topology in json and feed this to a an internal Golang API provided by ContainerLab. Would this be possible today or require quite a bit of work?

**Comment by hellt** [comment]
Hi @ldelossa 
yeah, this is a pain point at the moment.

Currently, the code base has some mixed spaghetti split between the `cmd` package and the rest of the backend stuff and it is the right time to address this.

Having said that, if your use case is as how you described it -> hand some json to containerlab via the exposed Go API and let it deploy it it might be possible.

I am connecting @steiler who once wanted to dive into this, maybe we can start working on this, and if you describe your high-level workflow with a bit more details we can check if that would make an MVP

**Reply by ldelossa** [reply]
Thanks a ton @hellt

**Comment by steiler** [comment]
Hi,
yes, I've identified that point quite some time ago. So we do unfortunately not have a clear and lean facade for containerlab, as Roman mentioned, deployment logic is scattered over the cmd and clab package.
I have now started a work in progress PR #1906. I might come up with a couple of questions on what you guys think is the proper / best way of doing things. So please join me there by discuss how to do it the right way and also review stuff.

**Reply by ldelossa** [reply]
Great. 

I do not want to hijack that issue with Cilium related requirements for such a library interaction, unless you are saying this is the correct place to add those concerns?

**Reply by hellt** [reply]
this is the right place @ldelossa 
we have one other use case (c9s) that would benefit from consuming clab as a library, but having an outside project is helpful.
So if you know the requirements you want clab to fulfill that would help us design the API in a more uniformal way

**Comment by unikzforce** [comment]
i'm curious if it's still on the roadmap

**Reply by steiler** [reply]
Hi, yes it for sure is. I'm a little short on time atm. So it will still take some time, sorry.

**Reply by unikzforce** [reply]
Thanks a lot @steiler

**Comment by Lachstec** [comment]
Hi there, I am also curious if this is still planned and would volunteer on working on it, if help is wanted :)

**Reply by hellt** [reply]
yes, please chime in. Contributions are welcome

**Reply by Lachstec** [reply]
Perfect! Any feelings on how this should get implemented, based on the previous discussions? Maybe start by resolving the spaghetti you mentioned and then think about what should get exposed as library interface?

**Comment by hellt** [comment]
Please try Containerlab v0.70.0

The CLI has been decoupled from the core, you should be able to use Containerlab as a package now
